After upgrade, certain jobs are not terminating

**Describe the bug**

Certain jobs are not being terminated despite the output return exit 0. It impacts on the subsquent jobs. A rpm upgrade from 3.0.13 to 3.0.19 was made prior to this error. 

**My Rundeck detail**

* Rundeck version: UPgraded from 3.0.13 to 3.0.19
* install type: rpm update
* OS Name/version: Centos 7.5.1804 
* DB Type/version: mysql in Galera Cluster with Haproxy

**To Reproduce**
Steps to reproduce the behavior:

1. Have prior jobs scheduled
2. Upgrade via rpm to 3.0.19.
3.  Restart Rundeck



**Expected behavior**
Jobs to finish with status succeeded.

**Screenshots**
If applicable, add screenshots to help explain your problem.
<img width="826" alt="Capture-error" src="https://user-images.githubusercontent.com/18469770/55422549-3ff1c800-5552-11e9-865d-3184b663c7fb.PNG">


[error-deadlock.txt](https://github.com/rundeck/rundeck/files/3035198/error-deadlock.txt)


**Additional context**

The upgrade was made with an attempt to reduce the deadlock database errors.

[profile.txt](https://github.com/rundeck/rundeck/files/3035182/profile.txt)

[rundeck-config.properties.txt](https://github.com/rundeck/rundeck/files/3035183/rundeck-config.properties.txt)

[framework-acme.properties.txt](https://github.com/rundeck/rundeck/files/3035189/framework-acme.properties.txt)

Hello @ProTip 

Got same issue with long running jobs.

Tried:
```
rundeck.disable.ref.stats = true
```
with
```
dataSource.url=jdbc:mysql://127.0.0.1/rundeck?autoReconnect=true&useSSL=false
...
dataSource.properties.maxActive=200
dataSource.properties.testOnBorrow=true
dataSource.properties.testWhileIdle=true
dataSource.properties.testOnReturn=true
```
But still no luck - all nodes processed but job is stuck in running state.

Here is my logs:
```
2019-04-03 17:32:29.183 ERROR --- [eduler_Worker-6] g.p.q.l.ExceptionPrinterJobListener      : Exception occurred in job: null

org.quartz.JobExecutionException: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: org.springframework.transaction.CannotCreateTransactionException: Could not open Hibernat
e Session for transaction; nested exception is org.hibernate.TransactionException: JDBC begin transaction failed: ]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:218)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.quartz.SchedulerException: Job threw an unhandled exception.
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        ... 1 common frames omitted
Caused by: org.springframework.transaction.CannotCreateTransactionException: Could not open Hibernate Session for transaction; nested exception is org.hibernate.TransactionException: JDBC begin transaction faile
d: 
        at org.springframework.orm.hibernate5.HibernateTransactionManager.doBegin(HibernateTransactionManager.java:542)
        at org.grails.orm.hibernate.GrailsHibernateTransactionManager.doBegin(GrailsHibernateTransactionManager.groovy:66)
        at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:377)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:130)
        at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
        at rundeck.services.ExecutionService.isApplicationShutdown(ExecutionService.groovy)
        at sun.reflect.GeneratedMethodAccessor1086.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1225)
        at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1125)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1092)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
        at groovy.lang.Closure.call(Closure.java:418)
        at groovy.lang.Closure.call(Closure.java:412)
        at java_util_concurrent_Callable$call.call(Unknown Source)

      at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callSafe(AbstractCallSite.java:83)
        at rundeck.quartzjobs.ExecutionJob.withRetry(ExecutionJob.groovy:490)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrap.invoke(PogoMetaMethodSite.java:190)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
        at rundeck.quartzjobs.ExecutionJob.saveState(ExecutionJob.groovy:582)
        at rundeck.quartzjobs.ExecutionJob$saveState$5.callCurrent(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
        at rundeck.quartzjobs.ExecutionJob$saveState$5.callCurrent(Unknown Source)
        at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:171)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:352)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
        at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:68)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:169)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:98)
        at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
        at groovy.lang.Closure.call(Closure.java:418)
        at groovy.lang.Closure.call(Closure.java:412)
        at com.codahale.metrics.Timer.time(Timer.java:99)
        at com.codahale.metrics.Timer$time$0.call(Unknown Source)
        at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:97)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 common frames omitted
Caused by: org.hibernate.TransactionException: JDBC begin transaction failed: 
        at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.begin(AbstractLogicalConnectionImplementor.java:72)
        at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.begin(LogicalConnectionManagedImpl.java:235)
        at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.begin(JdbcResourceLocalTransactionCoordinatorImpl.java:214)
        at org.hibernate.engine.transaction.internal.TransactionImpl.begin(TransactionImpl.java:52)
        at org.hibernate.internal.SessionImpl.beginTransaction(SessionImpl.java:1525)
        at org.springframework.orm.hibernate5.HibernateTransactionManager.doBegin(HibernateTransactionManager.java:500)
        ... 57 common frames omitted
Caused by: com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: Communications link failure

The last packet successfully received from the server was 1,889,742 milliseconds ago.  The last packet sent successfully to the server was 61 milliseconds ago.
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
        at com.mysql.jdbc.SQLError.createCommunicationsException(SQLError.java:990)
        at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2518)
        at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2683)
        at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2491)
        at com.mysql.jdbc.ConnectionImpl.setAutoCommit(ConnectionImpl.java:4807)
        at sun.reflect.GeneratedMethodAccessor206.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.tomcat.jdbc.pool.ProxyConnection.invoke(ProxyConnection.java:126)
        at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
        at org.apache.tomcat.jdbc.pool.interceptor.ConnectionState.invoke(ConnectionState.java:152)
        at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
        at org.apache.tomcat.jdbc.pool.TrapException.invoke(TrapException.java:40)
        at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
        at org.apache.tomcat.jdbc.pool.interceptor.AbstractCreateStatementInterceptor.invoke(AbstractCreateStatementInterceptor.java:79)
        at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
        at org.apache.tomcat.jdbc.pool.DisposableConnectionFacade.invoke(DisposableConnectionFacade.java:81)
        at com.sun.proxy.$Proxy75.setAutoCommit(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor206.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy$LazyConnectionInvocationHandler.invoke(LazyConnectionDataSourceProxy.java:376)
        at com.sun.proxy.$Proxy77.setAutoCommit(Unknown Source)
```

@ProTip
Just tried with:
```
dataSource.url=jdbc:mysql://10.1.1.102/tower?autoReconnectForPools=true&useSSL=false
dataSource.pooled=true
dataSource.properties.removeAbandoned=true
dataSource.properties.removeAbandonedTimeout=200
dataSource.properties.testOnBorrow=true
dataSource.properties.testWhileIdle=true
dataSource.properties.testOnReturn=true
```
Result still the same - hung in running state after processing all nodes.

Here is my logs:
```
2019-04-03 18:26:25.511 ERROR --- [eduler_Worker-1] g.p.q.l.ExceptionPrinterJobListener      : Exception occurred in job: null

org.quartz.JobExecutionException: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: org.springframework.transaction.CannotCreateTransactionException: Could not open Hibernate Session for transaction; nested exception is org.hibernate.TransactionException: JDBC begin transaction failed: ]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:218)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: org.quartz.SchedulerException: Job threw an unhandled exception.
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	... 1 common frames omitted
Caused by: org.springframework.transaction.CannotCreateTransactionException: Could not open Hibernate Session for transaction; nested exception is org.hibernate.TransactionException: JDBC begin transaction failed: 
	at org.springframework.orm.hibernate5.HibernateTransactionManager.doBegin(HibernateTransactionManager.java:542)
	at org.grails.orm.hibernate.GrailsHibernateTransactionManager.doBegin(GrailsHibernateTransactionManager.groovy:66)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:377)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:130)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at rundeck.services.ExecutionService.isApplicationShutdown(ExecutionService.groovy)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1225)
	at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1125)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1092)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
	at groovy.lang.Closure.call(Closure.java:418)
	at groovy.lang.Closure.call(Closure.java:412)
	at java_util_concurrent_Callable$call.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callSafe(AbstractCallSite.java:83)
	at rundeck.quartzjobs.ExecutionJob.withRetry(ExecutionJob.groovy:490)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrap.invoke(PogoMetaMethodSite.java:190)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:157)
	at rundeck.quartzjobs.ExecutionJob.saveState(ExecutionJob.groovy:582)
	at rundeck.quartzjobs.ExecutionJob$saveState$5.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:157)
	at rundeck.quartzjobs.ExecutionJob.execute_internal(ExecutionJob.groovy:171)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:352)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:68)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:157)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:169)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy:98)
	at rundeck.quartzjobs.ExecutionJob$_execute_closure1.doCall(ExecutionJob.groovy)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
	at groovy.lang.Closure.call(Closure.java:418)
	at groovy.lang.Closure.call(Closure.java:412)
	at com.codahale.metrics.Timer.time(Timer.java:99)
	at com.codahale.metrics.Timer$time$0.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
	at rundeck.quartzjobs.ExecutionJob.execute(ExecutionJob.groovy:97)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	... 1 common frames omitted
Caused by: org.hibernate.TransactionException: JDBC begin transaction failed: 
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.begin(AbstractLogicalConnectionImplementor.java:72)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.begin(LogicalConnectionManagedImpl.java:235)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.begin(JdbcResourceLocalTransactionCoordinatorImpl.java:214)
	at org.hibernate.engine.transaction.internal.TransactionImpl.begin(TransactionImpl.java:52)
	at org.hibernate.internal.SessionImpl.beginTransaction(SessionImpl.java:1525)
	at org.springframework.orm.hibernate5.HibernateTransactionManager.doBegin(HibernateTransactionManager.java:500)
	... 65 common frames omitted
Caused by: java.sql.SQLException: Connection has already been closed.
	at org.apache.tomcat.jdbc.pool.ProxyConnection.invoke(ProxyConnection.java:117)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
	at org.apache.tomcat.jdbc.pool.interceptor.ConnectionState.invoke(ConnectionState.java:152)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
	at org.apache.tomcat.jdbc.pool.TrapException.invoke(TrapException.java:40)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
	at org.apache.tomcat.jdbc.pool.interceptor.AbstractCreateStatementInterceptor.invoke(AbstractCreateStatementInterceptor.java:79)
	at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:108)
	at org.apache.tomcat.jdbc.pool.DisposableConnectionFacade.invoke(DisposableConnectionFacade.java:81)
	at com.sun.proxy.$Proxy75.setAutoCommit(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor207.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy$LazyConnectionInvocationHandler.invoke(LazyConnectionDataSourceProxy.java:376)
	at com.sun.proxy.$Proxy77.setAutoCommit(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor207.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.jdbc.datasource.TransactionAwareDataSourceProxy$TransactionAwareInvocationHandler.invoke(TransactionAwareDataSourceProxy.java:240)
	at com.sun.proxy.$Proxy77.setAutoCommit(Unknown Source)
	at org.hibernate.resource.jdbc.internal.AbstractLogicalConnectionImplementor.begin(AbstractLogicalConnectionImplementor.java:67)
	... 70 common frames omitted
```

--- a/rundeckapp/grails-app/jobs/rundeck/quartzjobs/ExecutionJob.groovy
+++ b/rundeckapp/grails-app/jobs/rundeck/quartzjobs/ExecutionJob.groovy
@@ -21,14 +21,12 @@ import com.codahale.metrics.Timer
 import com.dtolabs.rundeck.core.authorization.AuthContext
 import com.dtolabs.rundeck.core.dispatcher.DataContextUtils
 import com.dtolabs.rundeck.core.dispatcher.ExecutionState
-import com.dtolabs.rundeck.core.execution.ServiceThreadBase
 import com.dtolabs.rundeck.core.execution.WorkflowExecutionServiceThread
 import grails.core.GrailsApplication
 import grails.gorm.transactions.Transactional
 import org.quartz.JobExecutionContext
 import com.dtolabs.rundeck.core.common.Framework
 import org.quartz.InterruptableJob
-
 import com.dtolabs.rundeck.core.execution.workflow.NodeRecorder
 import org.rundeck.util.Sizes
 import rundeck.Execution
@@ -234,7 +232,7 @@ class ExecutionJob implements InterruptableJob {
 
         if(initMap.isTemp){
             //an adhoc execution without associated job
-            initMap.execution = Execution.get(initMap.executionId)
+            initMap.execution = fetchExecution(initMap.executionId)
             if (!initMap.execution) {
                 throw new RuntimeException("failed to lookup Exception object from job data map: id: ${initMap.executionId}")
             }
@@ -250,7 +248,7 @@ class ExecutionJob implements InterruptableJob {
             initMap.executionId=jobDataMap.get("executionId")
             initMap.secureOpts=jobDataMap.get("secureOpts")
             initMap.secureOptsExposed=jobDataMap.get("secureOptsExposed")
-            initMap.execution = Execution.get(initMap.executionId)
+            initMap.execution = fetchExecution(initMap.executionId)
             //NOTE: Oracle/hibernate bug workaround: if session has not flushed we may have to wait until Execution.get
             //can return the right entity
             int retry=30
@@ -259,7 +257,7 @@ class ExecutionJob implements InterruptableJob {
             }
             while(!initMap.execution && retry>0){
                 Thread.sleep(2000)
-                initMap.execution = Execution.get(initMap.executionId)
+                initMap.execution = fetchExecution(initMap.executionId)
                 retry--;
             }
             if (!initMap.execution) {
@@ -271,7 +269,6 @@ class ExecutionJob implements InterruptableJob {
             if (! initMap.execution instanceof Execution) {
                 throw new RuntimeException("JobDataMap contained invalid Execution type: " + initMap.execution.getClass().getName())
             }
-            initMap.execution.refresh()
             def jobArguments=initMap.frameworkService.parseOptsFromString(initMap.execution?.argString)
             if (initMap.scheduledExecution?.timeout && initMap.scheduledExecution?.timeout.contains('${')) {
                 def timeout = DataContextUtils.replaceDataReferencesInString(initMap.scheduledExecution?.timeout,
@@ -379,7 +376,11 @@ class ExecutionJob implements InterruptableJob {
         def WorkflowExecutionServiceThread thread = execmap.thread
         def Consumer<Long> periodicCheck = execmap.periodicCheck
         def ThresholdValue threshold = execmap.threshold
-        def jobAverageDuration = execmap.scheduledExecution?execmap.scheduledExecution.averageDuration:0
+        def jobAverageDuration
+        ScheduledExecution.withTransaction {
+            jobAverageDuration = execmap.scheduledExecution?execmap.scheduledExecution.averageDuration:0
+        }
+
         def boolean avgNotificationSent = false
         def boolean stop=false
 
@@ -610,6 +611,16 @@ class ExecutionJob implements InterruptableJob {
 
         }
         return saveStateComplete
+    }
+
+    @Transactional
+    Execution fetchExecution(def id) {
+        def execution = Execution.get(id)
+        if (execution != null) {
+            execution.refresh()
+        }
+
+        return execution
     }
 
     def ScheduledExecution fetchScheduledExecution(def jobDataMap) {

--- a/rundeckapp/grails-app/services/rundeck/services/ExecutionService.groovy
+++ b/rundeckapp/grails-app/services/rundeck/services/ExecutionService.groovy
@@ -50,6 +50,7 @@ import com.dtolabs.rundeck.plugins.scm.JobChangeEvent
 import com.dtolabs.rundeck.server.authorization.AuthConstants
 import grails.events.EventPublisher
 import grails.events.annotation.Subscriber
+import grails.gorm.transactions.NotTransactional
 import grails.gorm.transactions.Transactional
 import grails.web.mapping.LinkGenerator
 import groovy.transform.ToString
@@ -3957,6 +3958,7 @@ class ExecutionService implements ApplicationContextAware, StepExecutor, NodeSte
      * @throws StepException
      */
     @Override
+    @NotTransactional
     NodeStepResult executeNodeStep(StepExecutionContext executionContext, NodeStepExecutionItem executionItem,
                                    INodeEntry node) throws NodeStepException {
         if (!(executionItem instanceof JobExecutionItem)) {

--- a/test/api/test-job-long-run.sh
+++ b/test/api/test-job-long-run.sh
@@ -1,0 +1,161 @@
+#!/bin/bash
+
+#test  Ensure jobs are not holding up database connections
+
+DIR=$(cd `dirname $0` && pwd)
+source $DIR/include.sh
+
+###
+# setup: create a new job and acquire the ID
+###
+
+# job exec
+args="sleep 7"
+
+project=$2
+if [ "" == "$2" ] ; then
+    project="test"
+fi
+
+#escape the string for xml
+xmlargs=$($XMLSTARLET esc "$args")
+xmlproj=$($XMLSTARLET esc "$project")
+
+#produce job.xml content corresponding to the dispatch request
+cat > $DIR/temp.out <<END
+<joblist>
+   <job>
+      <name>Long Run Job</name>
+      <group>api-test/job-run</group>
+      <uuid>db9a5f0d</uuid>
+      <description></description>
+      <loglevel>INFO</loglevel>
+      <context>
+          <project>$xmlproj</project>
+      </context>
+      <dispatch>
+        <threadcount>1</threadcount>
+        <keepgoing>true</keepgoing>
+      </dispatch>
+      <sequence>
+        <command>
+        <exec>$xmlargs</exec>
+        </command>
+      </sequence>
+   </job>
+   <job>
+      <name>Long Run Wrapper</name>
+      <group>api-test/job-run</group>
+      <uuid>r2d2</uuid>
+      <description/>
+      <executionEnabled>true</executionEnabled>
+      <loglevel>INFO</loglevel>
+       <context>
+          <project>$xmlproj</project>
+      </context>
+      <nodeFilterEditable>false</nodeFilterEditable>
+      <sequence keepgoing="false" strategy="node-first">
+         <command>
+             <jobref name="Long Run Job">
+               <uuid>db9a5f0d</uuid>
+             </jobref>
+         </command>
+      </sequence>
+   </job>
+</joblist>
+
+END
+
+# now submit req
+runurl="${APIURL}/project/$project/jobs/import"
+
+params="dupeOption=update"
+
+# specify the file for upload with curl, named "xmlBatch"
+ulopts="-F xmlBatch=@$DIR/temp.out"
+
+# get listing
+docurl $ulopts  ${runurl}?${params} > $DIR/curl.out
+if [ 0 != $? ] ; then
+    errorMsg "ERROR: failed query request"
+    exit 2
+fi
+
+$SHELL $SRC_DIR/api-test-success.sh $DIR/curl.out || exit 2
+
+#result will contain list of failed and succeeded jobs, in this
+#case there should only be 1 failed or 1 succeeded since we submit only 1
+
+succount=$($XMLSTARLET sel -T -t -v "/result/succeeded/@count" $DIR/curl.out)
+jobid=$($XMLSTARLET sel -T -t -v "/result/succeeded/job[1]/id" $DIR/curl.out)
+ref_jobid=$($XMLSTARLET sel -T -t -v "/result/succeeded/job[2]/id" $DIR/curl.out)
+if [ "2" != "$succount" -o "" == "$jobid" ] ; then
+    errorMsg  "Upload was not successful."
+    exit 1
+fi
+
+
+runJob() {
+    ###
+    # Run the chosen id, expect success message
+    ###
+
+    local jobid=${1:?Must supply job id}
+
+    echo "TEST: POST job/id/run should succeed"
+
+
+    # now submit req
+    runurl="${APIURL}/job/${jobid}/run"
+    params=""
+
+    # get listing
+    $CURL -H "$AUTHHEADER" -X POST ${runurl}?${params} > $DIR/curl.out || fail "failed request: ${runurl}"
+
+    $SHELL $SRC_DIR/api-test-success.sh $DIR/curl.out || exit 2
+
+    #get execid
+
+    execcount=$($XMLSTARLET sel -T -t -v "/result/executions/@count" $DIR/curl.out)
+    execid=$($XMLSTARLET sel -T -t -v "/result/executions/execution/@id" $DIR/curl.out)
+
+    if [ "1" == "${execcount}" -a "" != "${execid}" ] ; then
+        :
+    else
+        errorMsg "FAIL: expected run success message for execution id. (count: ${execcount}, id: ${execid})"
+        exit 2
+    fi
+
+    #wait for execution to complete
+
+    api_waitfor_execution $execid 3 || fail "Failed waiting for execution $execid to complete"
+
+    # test execution status
+    # 
+    runurl="${APIURL}/execution/${execid}"
+
+    params=""
+
+    # get listing
+    docurl ${runurl}?${params} > $DIR/curl.out
+    if [ 0 != $? ] ; then
+        errorMsg "ERROR: failed query request"
+        exit 2
+    fi
+
+    $SHELL $SRC_DIR/api-test-success.sh $DIR/curl.out || exit 2
+
+    #Check projects list
+    itemcount=$($XMLSTARLET sel -T -t -v "/result/executions/@count" $DIR/curl.out)
+    assert "1" "$itemcount" "execution count should be 1"
+    status=$($XMLSTARLET sel -T -t -v "/result/executions/execution/@status" $DIR/curl.out)
+    assert "succeeded" "$status" "execution status should be succeeded"
+
+    echo "OK"
+
+    rm $DIR/curl.out
+}
+
+runJob $jobid
+
+runJob $ref_jobid

--- a/test/api/test-job-long-run.sh
+++ b/test/api/test-job-long-run.sh
@@ -54,9 +54,14 @@ cat > $DIR/temp.out <<END
           <project>$xmlproj</project>
       </context>
       <nodeFilterEditable>false</nodeFilterEditable>
-      <sequence keepgoing="false" strategy="node-first">
+      <sequence keepgoing="false" strategy="parallel">
          <command>
              <jobref name="Long Run Job">
+               <uuid>db9a5f0d</uuid>
+             </jobref>
+         </command>
+         <command>
+             <jobref name="Long Run Job" nodeStep="true">
                <uuid>db9a5f0d</uuid>
              </jobref>
          </command>

--- a/test/docker/dockers/rundeck/scripts/start_rundeck.sh
+++ b/test/docker/dockers/rundeck/scripts/start_rundeck.sh
@@ -257,6 +257,8 @@ server.address=0.0.0.0
 grails.serverURL=${RUNDECK_URL}
 dataSource.dbCreate = update
 dataSource.url = jdbc:h2:file:/home/rundeck/server/data/grailsdb;MVCC=true
+dataSource.properties.removeAbandoned=true
+dataSource.properties.removeAbandonedTimeout=5
 
 # Pre Auth mode settings
 rundeck.security.authorization.preauthenticated.enabled=false

